File: core/weapons/weapon.js
import Ship from "../player/ships/ship.js";
import { isSeperatingAxes } from "../utils/collision.js";
import { createVerticesPath, drawVerticesPath, tranformVertices } from "../utils/vertices.js";
import { ctx } from "../world/canvas.js";
import { spatial } from "../world/spatialHash.js";
import { drawWrapped, toroidalDistance, worldToScreen } from "../world/utils.js";
import WeaponManager from "./manager.js";

import { shapes } from "./shapes.js";


export default class Weapon {
    constructor({ name, type, x, y, width, height, angle, ship, acceleration = 10, speed = 10, damage = 10, range = 1000, energyCost = 10, fireRate = 1, vertices = shapes[0], color = "red", img }, force) {
        this.name = name;
        this.type = type;
        this.acceleration = acceleration;
        this.speed = speed;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
        this.angle = angle;
        this.damage = damage;
        this.img = img;

        this.range = range;
        this.fireRate = fireRate;
        this.energyCost = energyCost;

        this.vertices = vertices;

        this.color = color;
        this.ship = ship;

        this.dampSpeed = 0.99;

        this.path2D = createVerticesPath(tranformVertices(this.vertices, 0, -this.height / 2, this.width, this.height, 0));

        this.active = true;

        if (!force) {
            ship.increaseHeat(this.energyCost);
            ship.setCoolDown(1 / this.fireRate);
        }
    }

    render() {
        drawWrapped({
            fn: () => {
                ctx.rotate(-this.angle);
                // ctx.shadowColor = this.color;
                // ctx.shadowBlur = 10;
                if (this.img) {
                    ctx.drawImage(this.img, -this.width / 2, -this.height / 2, this.width, this.height);
                } else {
                    drawVerticesPath(this.path2D, this.color);
                }
            }, x: this.x, y: this.y, margin: {
                width: 200, height: 200
            }
        })
    }

    destroy() {
        WeaponManager.destroy(this);
    }

    getVertices() {
        const world = worldToScreen(this.x, this.y);
        const vertices = tranformVertices(this.vertices, world.x, world.y, this.width, this.height, this.angle);

        return vertices;
    }

    static nearBy(weapon, vertices, x, y) {
        const object = spatial.query(x, y, Math.ceil(weapon.height) + 1);


        for (const element of object) {
            if ((element instanceof Ship && (weapon.ship === element || element.state === "dead")) || (element instanceof Weapon && (weapon.ship === element.ship || weapon === element))) continue;


            weapon.closeObject(element)
            if (isSeperatingAxes(element.getVertices(), vertices).collision) {
                if (element instanceof Ship) {
                    element.life = Math.max(0, element.life - weapon.damage);
                    if (element.life <= 0) {
                        element.destroy();
                        weapon.ship.killScore++;
                    }
                } else {
                    element.destroy();
                }

                weapon.acceleration *= 0.8;
                weapon.range *= 0.8;


                if (!(--weapon.penetration)) {
                    weapon.colide();
                    weapon.destroy();
                }
            }
        }
    }

    getNearPlayers(radius) {
        const objects = spatial.query(this.x, this.y, Math.ceil(radius) + 1);

        const players = [];

        for (const obj of objects) {
            if (obj instanceof Ship && obj !== this.ship) {
                const dist = toroidalDistance(this.x, this.y, obj.x, obj.y);
                if (dist <= radius) {
                    players.push(obj);
                }
            }
        }
        return players;
    }

    getNearWeapons(radius) {
        const objects = spatial.query(this.x, this.y, Math.ceil(radius) + 1);

        const weapons = [];

        for (const obj of objects) {
            if (obj instanceof Weapon && obj !== this) {
                const dist = toroidalDistance(this.x, this.y, obj.x, obj.y);
                if (dist <= radius) {
                    weapons.push(obj);
                }
            }
        }
        return weapons;
    }


    colliding() {
        Weapon.nearBy(this, this.getVertices(), this.x, this.y);
    }

    colide() { }
    travelEnd() { }
    closeObject() { }
}

File: void_front/game/core/weapons/weapon.py
from __future__ import annotations

from game.core.player.ships.ship import Ship
from game.core.utils.collision import isSeperatingAxes
from game.core.utils.vertices import createVerticesPath, tranformVertices
from game.core.world.canvas import blit_image, draw_polygon
from game.core.world.spatial_hash import spatial
from game.core.world.utils import worldToScreen
import numpy as np


class Weapon:
    def __init__(self, options, force=False):
        from game.core.weapons.shapes import shapes

        self.name = options.get("name")
        self.type = options.get("type")
        self.acceleration = options.get("acceleration", 10)
        self.speed = options.get("speed", 10)
        self.x = options.get("x")
        self.y = options.get("y")
        self.width = options.get("width")
        self.height = options.get("height")
        self.angle = options.get("angle")
        self.damage = options.get("damage", 10)
        self.img = options.get("img")
        self.range = options.get("range", 1000)
        self.fireRate = options.get("fireRate", 1)
        self.energyCost = options.get("energyCost", 10)
        self.vertices = options.get("vertices", shapes[0])
        self.color = options.get("color", "red")
        self.ship = options.get("ship")
        self.dampSpeed = 0.99
        self.path2D = createVerticesPath(tranformVertices(self.vertices, 0, -self.height / 2, self.width, self.height, 0))
        self.active = True
        if not force and self.ship is not None:
            self.ship.increaseHeat(self.energyCost)
            self.ship.setCoolDown(1 / self.fireRate)

    def render(self):
        if self.img is not None:
            blit_image(self.img, (self.x, self.y), size=(self.width, self.height), angle=self.angle)
        else:
            draw_polygon(self.getVertices(), color=self.color, width=0, alpha=255)

    def destroy(self):
        from game.core.weapons.manager import WeaponManager

        WeaponManager.destroy(self)

    def getVertices(self):
        screen = worldToScreen(self.x, self.y)
        return tranformVertices(self.vertices, screen["x"], screen["y"], self.width, self.height, self.angle)

    @staticmethod
    def nearBy(weapon, vertices, x, y):
        object_list = spatial.query(x, y, np.ceil(weapon.height / spatial.cellSize) + 1)
        for element in object_list:
            if isinstance(element, Ship) and (weapon.ship == element or getattr(element, "state", None) == "dead"):
                continue
            if isinstance(element, Weapon) and (weapon.ship == element.ship or weapon is element):
                continue
            if hasattr(weapon, "closeObject"):
                weapon.closeObject(element)
            if isSeperatingAxes(element.getVertices(), vertices).get("collision"):
                weapon.ship.damage_score = weapon.damage
                if isinstance(element, Ship):
                    element.life = max(0, element.life - weapon.damage)
                    if element.life <= 0:
                        element.destroy()
                        weapon.ship.killScore += 1
                else:
                    element.destroy()
                weapon.acceleration *= 0.8
                weapon.range *= 0.8
                if hasattr(weapon, "penetration"):
                    weapon.penetration -= 1
                    if not weapon.penetration:
                        weapon.colide()
                        weapon.destroy()

    def colliding(self):
        Weapon.nearBy(self, self.getVertices(), self.x, self.y)

    def colide(self):
        pass

    def travelEnd(self):
        pass

    def closeObject(self, obj):
        pass
